pkg/parse/parse.go

```diff
@@ -496,61 +531,100 @@ func GetCacheMount(args []string) (specs.Mount, error) {
 			}
 		case "uid":
 			uid, err = strconv.Atoi(kv[1])
 			if err != nil {
-				return newMount, errors.Wrapf(err, "Unable to parse cache uid")
+				return newMount, "", errors.Wrapf(err, "Unable to parse cache uid")
 			}
 		case "gid":
 			gid, err = strconv.Atoi(kv[1])
 			if err != nil {
-				return newMount, errors.Wrapf(err, "Unable to parse cache gid")
+				return newMount, "", errors.Wrapf(err, "Unable to parse cache gid")
 			}
 		default:
-			return newMount, errors.Wrapf(errBadMntOption, kv[0])
+			return newMount, "", errors.Wrapf(errBadMntOption, kv[0])
 		}
 	}
 
 	if !setDest {
-		return newMount, noDestError
+		return newMount, "", noDestError
 	}
 
-	// since type is cache and cache can be resused by consecutive builds
-	// create a common cache directory, which should persists on hosts within temp lifecycle
-	// add subdirectory if specified
-	if id != "" {
-		newMount.Source = filepath.Join(GetTempDir(), BuildahCacheDir, id)
+	if fromImage != "" {
+		// do not create cache on host
+		// instead use read-only mounted image as cache
+		mountPoint := ""
+		if additionalMountPoints != nil && !selinux.GetEnabled() {
+			if val, ok := additionalMountPoints[fromImage]; ok {
+				mountPoint = val
+			}
+		}
+		// if mountPoint of image was not found in additionalMap
+		// or additionalMap was nil, try mounting image
+		if mountPoint == "" {
+			image, err := util.LookupImage(store, fromImage)
+			if err != nil {
+				return newMount, "", err
+			}
+
+			mountPoint, err = image.Mount(context.Background(), nil, imageMountLabel)
+			if err != nil {
+				return newMount, "", err
+			}
+		}
+		contextDir := mountPoint
+		newMount.Options = append(newMount.Options, "ro")
+		if strings.HasPrefix(newMount.Source, ".") || newMount.Source == "" || !filepath.IsAbs(newMount.Source) || fromImage != "" {
+			// path should be /contextDir/specified path
+			newMount.Source = filepath.Join(contextDir, filepath.Clean(string(filepath.Separator)+newMount.Source))
+		}
 	} else {
-		newMount.Source = filepath.Join(GetTempDir(), BuildahCacheDir)
-	}
-	// create cache on host if not present
-	err = os.MkdirAll(newMount.Source, os.FileMode(mode))
-	if err != nil {
-		return newMount, errors.Wrapf(err, "Unable to create build cache directory")
-	}
-	//buidkit parity: change uid and gid if specificed otheriwise keep `0`
-	err = os.Chown(newMount.Source, uid, gid)
-	if err != nil {
-		return newMount, errors.Wrapf(err, "Unable to change uid,gid of cache directory")
+		// we need to create cache on host if no image is being used
+
+		// since type is cache and cache can be resused by consecutive builds
+		// create a common cache directory, which should persists on hosts within temp lifecycle
+		// add subdirectory if specified
+		if id != "" {
+			newMount.Source = filepath.Join(GetTempDir(), BuildahCacheDir, id)
+		} else {
+			newMount.Source = filepath.Join(GetTempDir(), BuildahCacheDir)
```

Comment:
`id` specifies a different `cache` directory on host if user wants. Content is `cached` on host it is expected to use a common `cache` directory if `id` is not specified.

For the case of `cache` with `from` , `id` is practically `noop` since src of a cache is a `image` or a `stage` not a cache on host machine so no point of differentiation.

Reply:
Fixed in latest commit.